util/hot_backup.cc

```diff
@@ -0,0 +1,654 @@
+    Version * Ver,
+    Options & BackupOptions)
+{
+    bool good(true);
+    int level, ret_val;
+    std::string src_path, dst_path;
+
+    for (level=0; level<config::kNumLevels && good; ++level)
+    {
+        const Version::FileMetaDataVector_t level_files(Ver->GetFileList(level));
+        Version::FileMetaDataVector_t::const_iterator it;
+
+        for (it=level_files.begin(); it!=level_files.end() && good; ++it)
+        {
+            src_path=TableFileName(options_, (*it)->number, level);
+            dst_path=TableFileName(BackupOptions, (*it)->number, level);
+
+            // this would be cleaner if link operation added to Env class heirarchy
+            ret_val=link(src_path.c_str(), dst_path.c_str());
+            if (0!=ret_val)
+            {
+                good=false;
+                Log(GetLogger(), "HotBackup failed linking (return val %d) %s",
+                    ret_val, src_path.c_str());
+            }   // if
+        }   // for
+    }   // for
+
+    return(good);
+
+}   // DBImpl::CreateBackupLinks
+
+
+/**
+ * Copy portion of LOG file that most likely applies to this backup.
+ */
+bool
+DBImpl::CopyLOGSegment(long EndPos)
+{
+    assert(0<EndPos);
+
+    bool good(true);
+    long remaining;
+    Options local_options;
+    SequentialFile * src(NULL);
+    WritableFile * dst(NULL);
+    Status s;
+    std::string src_name, dst_name, buffer;
+    Slice data_read;
+
+    remaining=EndPos;
+
+    // open source and destination LOG files
+    local_options=options_;
+    SetBackupPaths(local_options, 0);
+
+    src_name=InfoLogFileName(options_.tiered_fast_prefix);
+    dst_name=InfoLogFileName(local_options.tiered_fast_prefix);
+
+    s=options_.env->NewSequentialFile(src_name, &src);
+    if (s.ok())
+    {
+        s=options_.env->NewWritableFile(dst_name, &dst, 4096);
+
+        if (!s.ok())
+            Log(GetLogger(), "HotBackup failed to create destination LOG file (%s, %s)",
+                s.ToString().c_str(), dst_name.c_str());
+    }   // if
+    else
+    {
+        Log(GetLogger(), "HotBackup failed to open source LOG file (%s, %s)",
+            s.ToString().c_str(), src_name.c_str());
+    }   // else
+
+    // copy data between files
+    buffer.reserve(4096);
+
+    while(0<remaining && s.ok())
+    {
+        long count;
+
+        count=(4096<remaining ? 4096 : remaining);
+        s=src->Read(count, &data_read, (char *)buffer.data());
+
+        if (s.ok())
+        {
+            s=dst->Append(data_read);
+
+            if (!s.ok())
+                Log(GetLogger(), "HotBackup LOG write failed at %ld (%s)",
+                    remaining, s.ToString().c_str());
+        }   // if
+        else
+        {
+                Log(GetLogger(), "HotBackup LOG read failed at %ld (%s)",
+                    remaining, s.ToString().c_str());
+        }   // else
+
+        remaining-=count;
+    }   // while
+
+    // close files and go home
+    good=s.ok();
+    delete src;
```

Comment:
My issue is that the open file calls require a pointer to the pointer.  I would need to have a simple pointer to pass into the open file calls.  Then shove it into the auto_ptr ... or rewrite the open file calls to take an auto_ptr.  Not in the mood for either at this time.  auto_ptr is better code, but feels like we would be taking a hammer to a screw.